### PR TITLE
[compiler] Option to always take the non-memo branch

### PR DESCRIPTION
**Pull request by mvitousek**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #29655

Summary: This adds a debugging mode to the compiler that simply adds a `|| true` to the guard on all memoization blocks, which results in the generated code never using memoized values and always recomputing them. This is designed as a validation tool for the compiler's correctness--every program *should* behave exactly the same with this option enabled as it would with it disabled, and so any difference in behavior should be investigated as either a compiler bug or a pipeline issue.

(We add `|| true` rather than dropping the conditional block entirely because we still want to exercise the guard tests, in case the guards themselves are the source of an error, like reading a property from undefined in a guard.)

### CODE REVIEW

File: compiler/packages/snap/src/compiler.ts [hunk 1]
```diff
@@ -124,6 +125,9 @@ function makePluginOptions(
   if (firstLine.includes("@enablePreserveExistingManualUseMemo")) {
     enablePreserveExistingManualUseMemo = true;
   }
+  if (firstLine.includes("@disableMemoizationForDebugging")) {
+    disableMemoizationForDebugging = true;
+  }
```
Comment (josephsavona) [inline compiler/packages/snap/src/compiler.ts:130]: same as previous PR, this should get picked up by the call to parseConfigPragmaFn

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/useMemo-simple-preserved-nomemo.expect.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  } else {
+    t1 = $[1];
+  }
+  let t2;
+  if ($[2] === Symbol.for("react.memo_cache_sentinel")) {
+    t2 = [];
+    $[2] = t2;
+  } else {
+    t2 = $[2];
+  }
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/useMemo-simple-preserved-nomemo.expect.md:44]: is it intentional that scopes w/o deps don't get the `|| true`?

**Review by josephsavona — COMMENTED** [review]
This is an awesome idea for debugging! I love that this has already helped to identify some issues.

One question though: it seems like it would be more consistent for this mode to never memoize any values that weren't explicitly (manually) memoized in the original code. As implemented, we are actually memoizing values w/o any deps.

**Comment by mvitousek** [comment]
Good catch! Fixed.

**Review by josephsavona — APPROVED** [review]
awesomesauce

**Pull request merged**